components/script/dom/htmlmeterelement.rs

```diff
@@ -56,6 +76,86 @@ impl HTMLMeterElement {
             can_gc,
         )
     }
+
+    fn create_shadow_tree(&self, can_gc: CanGc) {
+        let document = self.owner_document();
+        let root = self
+            .upcast::<Element>()
+            .attach_shadow(
+                IsUserAgentWidget::Yes,
+                ShadowRootMode::Closed,
+                false,
+                SlotAssignmentMode::Manual,
+            )
+            .expect("Attaching UA shadow root failed");
+
+        let meter_value = HTMLDivElement::new(local_name!("div"), None, &document, None, can_gc);
+        root.upcast::<Node>()
+            .AppendChild(meter_value.upcast::<Node>())
+            .unwrap();
+
+        let _ = self.shadow_tree.borrow_mut().insert(ShadowTree {
+            meter_value: meter_value.as_traced(),
+        });
+        self.upcast::<Node>()
+            .dirty(crate::dom::node::NodeDamage::OtherNodeDamage);
+    }
+
+    fn shadow_tree(&self, can_gc: CanGc) -> Ref<'_, ShadowTree> {
+        if !self.upcast::<Element>().is_shadow_host() {
+            self.create_shadow_tree(can_gc);
+        }
+
+        Ref::filter_map(self.shadow_tree.borrow(), Option::as_ref)
+            .ok()
+            .expect("UA shadow tree was not created")
+    }
+
+    fn update_state(&self, can_gc: CanGc) {
+        let value = *self.Value();
+        let low = *self.Low();
+        let high = *self.High();
+        let min = *self.Min();
+        let max = *self.Max();
+        let optimum = *self.Optimum();
+
+        // Compute pseudo class
```

Comment:
It would be good to add documentation as to what steps of the spec this implements. 

The below seems to correspond to various prose in "UA requirements for regions of the gauge"(no link unfortunately), so the various conditions could be document with that exact same prose. 

(the same could be done for various methods like `self.Value()`, but that is outside of the scope of this PR)

Reply:
Yeah makes sense - I initially didn't do this because it's a lot of text :D